### PR TITLE
Prevent Archetypes::get_2_mut from being called with two empty archetype ids

### LINKED ISSUE #2435: spawn_batch using empty bundle panics
Bevy version: b52edc1

```rs
use bevy::ecs::prelude::*;

fn main() {
    let mut world = World::new();
    world.spawn_batch((0..1).map(|_| ()));
}
```
This code panics.
## Additional information

Initially found on discord: https://discord.com/channels/691052431525675048/742884593551802431/861612284945891339

### PR DESCRIPTION
**Pull request by msvbg**
# Objective

Emboldened by my last PR, I decided to try fixing another ECS bug.

Closes #2435.

## Solution

~~This PR changes the `Archetype::get_2_mut` function to represent a potential footgun, passing in the same id twice, in the type system. I noticed that there is an equivalent function in `Table::get_2_mut` which does the same thing, so I changed it too even though it doesn't seem to be misused in the same way.~~

This PR now only adds an if clause to check if `archetype_id` is the empty archetype, and treats it as a special case.

### CODE REVIEW

File: crates/bevy_ecs/src/archetype.rs [hunk 1]
```diff
@@ -447,18 +447,24 @@ impl Archetypes {
...
     pub(crate) fn get_2_mut(
         &mut self,
         a: ArchetypeId,
         b: ArchetypeId,
-    ) -> (&mut Archetype, &mut Archetype) {
+    ) -> (&mut Archetype, Option<&mut Archetype>) {
+        if a == b {
+            return (self.archetypes.get_mut(a.index()).unwrap(), None);
+        }
         if a.index() > b.index() {
```
Comment (MinerSebas) [inline crates/bevy_ecs/src/archetype.rs:462]:
```suggestion
            (self.archetypes.get_mut(a.index()).unwrap(), None)
        }
        else if a.index() > b.index() {
```

File: crates/bevy_ecs/src/storage/table.rs [hunk 2]
```diff
@@ -439,14 +439,20 @@ impl Tables {
...
+    /// Given two `TableId`s, returns two mutable references corresponding
+    /// to those tables. If two identical ids are provided, the second tuple
+    /// element is `None`.
     #[inline]
-    pub(crate) fn get_2_mut(&mut self, a: TableId, b: TableId) -> (&mut Table, &mut Table) {
+    pub(crate) fn get_2_mut(&mut self, a: TableId, b: TableId) -> (&mut Table, Option<&mut Table>) {
+        if a == b {
+            return (self.tables.get_mut(a.index()).unwrap(), None);
+        }
         if a.index() > b.index() {
```
Comment (MinerSebas) [inline crates/bevy_ecs/src/storage/table.rs:450]:
```suggestion
            (self.tables.get_mut(a.index()).unwrap(), None)
        }
        else if a.index() > b.index() {
```

**Comment by cart** [comment]
Relying on the "index out of bounds" panic here was an intentional choice. These methods shouldn't ever be called with identical indices. This is performance critical code and I don't want to add extra Options where they aren't absolutely needed, as "unwrapping" them adds a branch. Additionally, as users should never call this with identical indices, there is no "valid" instance where they should need to handle this case. Handling this complexity "internally" instead of pushing it to the user is, imo, the right call.

**Comment by msvbg** [comment]
I see your point. I'll try to rework the PR to avoid Options and just fix the issue at the call site.

**Comment by msvbg** [comment]
Done. The easiest way to please the borrow checker was to duplicate some code.

**Review by alice-i-cecile — APPROVED** [review]

**Comment by msvbg** [comment]
I think this PR is obsolete now, so I'm closing it.

**Pull request closed without merging**